packages/cosmic-swingset/src/launch-chain.js

```diff
@@ -586,21 +586,26 @@ export async function launch({
     return p;
   }
 
-  function checkExpectedBlockHeight(blockHeight) {
-    if (computedHeight > 0 && computedHeight !== blockHeight) {
-      // We only tolerate the trivial case.
-      const restoreHeight = blockHeight - 1;
-      if (restoreHeight !== computedHeight) {
-        // Keep throwing forever.
-        decohered = Error(
-          // TODO unimplemented
-          `Unimplemented reset state from ${computedHeight} to ${restoreHeight}`,
-        );
-        throw decohered;
-      }
+  function blockNeedsExecution(blockHeight) {
+    if (blockHeight > 0 && savedHeight === blockHeight) {
```

Comment:
wait, shouldn't that be `if (savedHeight > 0` ?

Would it be correct for this to read:

```js
if (savedHeight === 0) {
  // 0 is the default we use when the DB is empty, so either we haven't executed any blocks, or we've executed the bootstrap block but no others. The first non-bootstrap block can have an arbitrary height (the chain may not start at 1), so we can't tell if it is a replay or not, so assume the block needs to be executed
  return true;
}

if (blockHeight === savedHeight + 1) {
  // execute the next block
  return true;
}

if (blockHeight === savedHeight) {
  // we have already committed this block, so "replay" by not executing (but returning all the results from the last time)
  return false;
}

// we're being asked to rewind by more than one block, or execute something more than one block in the future, neither of which we can accomodate
decohered = ...
throw decohered;
```